File: Data Structures Intro/Linked List/linked-list-task-net/CustomLinkedList/CustomLinkedList.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CustomLinkedList
{
    /// <summary>
    /// Basic data structure for custom Linked List representation
    /// </summary>
    public class CustomLinkedList<T>
    {
        /// <summary>
        /// Internal class for representing a linked list node
        /// </summary>
        protected class CustomListNode
        {
            public T Element { get; set; }
            public CustomListNode? Next { get; set; }

            public CustomListNode(T element, CustomListNode? next)
            {
                Element = element;
                Next = next;
            }
        }

        /// <summary>
        /// Linked List head
        /// </summary>
        protected CustomListNode? Head { get; set; }

        /// <summary>
        /// Constructor from a list instance.
        /// </summary>
        /// <param name="elements">Elements that should be inserted into the result linked list.</param>
        public CustomLinkedList(IEnumerable<T> elements)
        {
            Head = null;
            CustomListNode? current = null;
            foreach (var element in elements)
            {
                var newNode = new CustomListNode(element, null);
                if (Head == null)
                {
                    Head = newNode;
                    current = Head;
                }
                else
                {
                    current!.Next = newNode;
                    current = newNode;
                }
            }
        }

        /// <summary>
        /// Removes all nodes containing the given element
        /// </summary>
        public CustomLinkedList<T> RemoveNodes(T element)
        {
            // Remove matching nodes from the head first
            while (Head != null && EqualityComparer<T>.Default.Equals(Head.Element, element))
            {
                Head = Head.Next;
            }

            var current = Head;
            while (current != null && current.Next != null)
            {
                if (EqualityComparer<T>.Default.Equals(current.Next.Element, element))
                {
                    current.Next = current.Next.Next;
                }
                else
                {
                    current = current.Next;
                }
            }
            return this;
        }

        /// <summary>
        /// Reverses the linked list in-place
        /// </summary>
        public CustomLinkedList<T> Reverse()
        {
            CustomListNode? prev = null;
            var current = Head;

            while (current != null)
            {
                var next = current.Next;
                current.Next = prev;
                prev = current;
                current = next;
            }

            Head = prev;
            return this;
        }

        /// <summary>
        /// Returns the list starting from the right middle node.
        /// If even number of nodes, choose the second middle.
        /// </summary>
        public CustomLinkedList<T> GetRightMiddle()
        {
            if (Head == null)
                return new CustomLinkedList<T>(Enumerable.Empty<T>());

            var slow = Head;
            var fast = Head;

            // fast moves twice as fast as slow
            while (fast != null && fast.Next != null)
            {
                slow = slow!.Next;
                fast = fast.Next.Next;
            }

            // Build new linked list from slow to the end
            var resultElements = new List<T>();
            var current = slow;
            while (current != null)
            {
                resultElements.Add(current.Element);
                current = current.Next;
            }

            return new CustomLinkedList<T>(resultElements);
        }

        /// <summary>
        /// Returns whether `this` instance is identical to `list` one, order of elements matters
        /// </summary>
        public bool Check(IEnumerable<T> elements)
        {
            if (Head == null)
                return !elements.Any();

            var curr = Head;
            foreach (var element in elements)
            {
                if (curr == null)
                    return false;
                if (!EqualityComparer<T>.Default.Equals(element, curr.Element))
                    return false;
                curr = curr.Next;
            }
            return curr == null;
        }
    }
}

File: Data Structures Intro/Linked List/Java/linked-list-task-java/src/main/java/com/epam/bsp/linked/list/CustomLinkedList.java
package com.epam.bsp.linked.list;

import java.util.List;

public class CustomLinkedList<E> {

    protected static class CustomListNode<E> {
        protected E element;
        protected CustomListNode<E> next;

        public CustomListNode(E element, CustomListNode<E> next) {
            this.element = element;
            this.next = next;
        }
    }

    protected CustomListNode<E> head;

    /**
     * Constructor: create linked list from given list of elements
     */
    public CustomLinkedList(List<E> elements) {
        head = null;
        if (elements == null || elements.isEmpty()) return;

        CustomListNode<E> current = null;
        for (E e : elements) {
            CustomListNode<E> newNode = new CustomListNode<>(e, null);
            if (head == null) {
                head = newNode;
                current = head;
            } else {
                current.next = newNode;
                current = newNode;
            }
        }
    }

    /**
     * Removes all nodes that contain given element
     */
    public CustomLinkedList<E> removeNodes(E element) {
        // Remove from head while needed
        while (head != null && (head.element == null ? element == null : head.element.equals(element))) {
            head = head.next;
        }

        CustomListNode<E> current = head;
        while (current != null && current.next != null) {
            if (current.next.element == null ? element == null : current.next.element.equals(element)) {
                current.next = current.next.next;
            } else {
                current = current.next;
            }
        }
        return this;
    }

    /**
     * Reverses the linked list
     */
    public CustomLinkedList<E> reverse() {
        CustomListNode<E> prev = null;
        CustomListNode<E> curr = head;
        while (curr != null) {
            CustomListNode<E> next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        head = prev;
        return this;
    }

    /**
     * Returns the linked list starting from the middle (if two middles, use second)
     */
    public CustomLinkedList<E> getRightMiddle() {
        if (head == null) return this;

        CustomListNode<E> slow = head;
        CustomListNode<E> fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        CustomLinkedList<E> result = new CustomLinkedList<>(List.of());
        result.head = slow;
        return result;
    }

    /**
     * Checks if linked list elements equal to given list elements (order matters)
     */
    public boolean check(List<E> elements) {
        CustomListNode<E> curr = head;
        for (E element : elements) {
            if (curr == null) return false;
            if (!(element == null ? curr.element == null : element.equals(curr.element))) {
                return false;
            }
            curr = curr.next;
        }
        return curr == null;
    }
}
